### PR TITLE
API: rename MultiIndex.labels to MultiIndex.codes

### LINKED ISSUE #13443: API:  MultiIndex.labels -> codes
The boat may have long sailed on this, but just for consideration.

I semi-frequently get the `.levels` and `.labels` of a `MultiIndex` backwards.  Maybe it's just me, but I think `labels` is the culprit, because in other pandas contexts, "labels" refer to the actual value of the thing.  E.g.
   - `.loc` indexes by "labels" (values)
   -  the values inside of a single row index or columns are the "labels" for those items

So, consistent with `Categorical`, would it make sense for the integer mapping inside a `MultiIndex` to also be called `.codes`?

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #13443
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This PR renames ``MultiIndex.labels`` to ``MultiIndex.codes``.

Also note that I've not changed the repr method or the repr output in various doc strings examples in anticipation of #22511 (this would be unneeded extra work). This should all be ok when #22511 is merged.

I would appreciate comments on the changes quickly, as this is a very large PR that touches on many parts of pandas, so it quickly gets out of sync.

@jorisvandenbossche?

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1010,6 +1010,14 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 
+- :attr:`MultiIndex.labels` has been deprecated and replaced by :attr:`MultiIndex.codes`.
+  The functionality is unchanged. This new name better reflects the natures of
+  these codes and makes the API more similar to the API for
+  :class:`CategoricalIndex`(:issue:`13443`).
+  As a concequence, other uses of the name ``labels`` have also been deprecated in ``MultiIndex`` and replaced with ``codes``:
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.rst]: concequence --> consequence

File: doc/source/whatsnew/v0.24.0.rst [hunk 2]
```diff
@@ -1010,6 +1010,14 @@ Other API Changes
...
+  - You should initialize a MultiIndex instance using a parameter named ``codes`` rather than ``labels``.
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.rst]: Double back-ticks around `MultiIndex`

File: pandas/tests/indexes/multi/test_constructor.py [hunk 3]
```diff
@@ -58,51 +58,60 @@ def test_constructor_nonhashable_names():
...
         idx.copy().set_levels([['a'], ['b']])
 
     with pytest.raises(ValueError, match=label_error):
-        idx.copy().set_labels([[0, 0, 0, 0], [0, 0]])
+        idx.copy().set_codes([[0, 0, 0, 0], [0, 0]])
+
+
+def test_labels_deprecated(idx):
+    # GH23752
+    with tm.assert_produces_warning(FutureWarning):
```
Comment (gfyoung) [inline pandas/tests/indexes/multi/test_constructor.py]: Why is there a `FutureWarning` here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py [hunk 4]
```diff
@@ -137,7 +138,12 @@ class MultiIndex(Index):
     ----------
     levels : sequence of arrays
         The unique labels for each level
+    codes : sequence of arrays
+        Integers for each level designating which label at each location
     labels : sequence of arrays
+        .. deprecated:: 0.24.0
+            Use ``codes`` instead
```
[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/indexes/multi.py]: Yeah, changed.

**Comment by topper-123** [comment]
I've updated now. The Travis failure is  unrelated (it's a unrelated test that consumes too much memory):

```
Worker 'gw1' crashed while running 'pandas/tests/series/test_rank.py::test_pct_max_many_rows'
```

I've also caught/done a few more renames, see latest commit. 

I'll start up updating the docs tonight or tomorrow. I'll do the doc changes in a seperate PR, as the current PR already is very large.

**Comment by topper-123** [comment]
The CircleCi failure is unrelated to this PR:

```
___________________________ test_tick_add_sub[Micro] ___________________________

cls = <class 'pandas.tseries.offsets.Micro'>

    @pytest.mark.parametrize('cls', tick_classes)
>   @example(n=2, m=3)
    @example(n=800, m=300)
    @example(n=1000, m=5)
    @given(n=st.integers(-999, 999), m=st.integers(-999, 999))
    def test_tick_add_sub(cls, n, m):

pandas/tests/tseries/offsets/test_ticks.py:41: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/conda/envs/pandas/lib/python3.6/site-packages/hypothesis/core.py:552: in execute
    ) % (test.__name__, text_repr[0],))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hypothesis.core.StateForActualGivenExecution object at 0x7fce0bb6db00>
message = 'Hypothesis test_tick_add_sub(cls=Micro, n=30, m=359) produces unreliable results: Falsified on the first call but did not on a subsequent one'

    def __flaky(self, message):
        if len(self.falsifying_examples) <= 1:
>           raise Flaky(message)
E           hypothesis.errors.Flaky: Hypothesis test_tick_add_sub(cls=Micro, n=30, m=359) produces unreliable results: Falsified on the first call but did not on a subsequent one

/opt/conda/envs/pandas/lib/python3.6/site-packages/hypothesis/core.py:727: Flaky
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_tick_add_sub(cls=Micro, n=30, m=359)
Unreliable test timings! On an initial run, this test took 1040.15ms, which exceeded the deadline of 500.00ms, but on a subsequent run it took 0.08 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None.
```

**Comment by topper-123** [comment]
Hey, is this ok?

This touches a lot of things, so very quickly gets into conflicts...

**Review by gfyoung — APPROVED** [review]
Over to you @jreback

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -572,54 +580,70 @@ def set_levels(self, levels, level=None, inplace=False,
...
+
+    @deprecate_kwarg(old_arg_name='labels', new_arg_name='codes')
+    def set_codes(self, codes, level=None, inplace=False,
+                  verify_integrity=True):
         """
-        Set new labels on MultiIndex. Defaults to returning
+        Set new codes on MultiIndex. Defaults to returning
         new index.
 
         Parameters
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can you add a versinoadded tag here

**Review by jreback — CHANGES_REQUESTED** [review]
tiny change. mainly want you to merge master. check if anything in the .rst needs to change here as well. ping on green.

**Comment by topper-123** [comment]
This azure checks ar stalled, but looks ok to me: It [reports the bug below](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=4664)

```
2018-12-04T07:41:21.1391588Z ##[error]./asv_bench/benchmarks/frame_methods.py(2,1): error F401: 'warnings' imported but unused
```

My Pr doesn't touch frame_methods.py, os I think this is a case of the new azure tests being too eager.

I'm make a seperate PR remove the ``import warnings`` line from frame_methods.py

**Comment by jreback** [comment]
> My Pr doesn't touch frame_methods.py, os I think this is a case of the new azure tests being too eager.

not its not too eager, its an issue with the asv's that was not being properly caught before. wait a little and will let you know when to rebase.

**Comment by topper-123** [comment]
Rebased & green.

File: doc/source/whatsnew/v0.24.0.rst [hunk 6]
```diff
@@ -1100,6 +1100,14 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 
+- :attr:`MultiIndex.labels` has been deprecated and replaced by :attr:`MultiIndex.codes`.
+  The functionality is unchanged. The new name better reflects the natures of
+  these codes and makes the ``MultiIndex`` API more similar to the API for
+  :class:`CategoricalIndex`(:issue:`13443`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: this might be a doc-rendering problem, but can resolve later

File: doc/source/whatsnew/v0.24.0.rst [hunk 7]
```diff
@@ -1100,6 +1100,14 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 
+- :attr:`MultiIndex.labels` has been deprecated and replaced by :attr:`MultiIndex.codes`.
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: i don't this is valid any longer, you can remove the attr

File: doc/source/whatsnew/v0.24.0.rst [hunk 8]
```diff
@@ -1100,6 +1100,14 @@ Other API Changes
...
+  The functionality is unchanged. The new name better reflects the natures of
+  these codes and makes the ``MultiIndex`` API more similar to the API for
+  :class:`CategoricalIndex`(:issue:`13443`).
+  As a consequence, other uses of the name ``labels`` in ``MultiIndex`` have also been deprecated and replaced with ``codes``:
+  - You should initialize a ``MultiIndex`` instance using a parameter named ``codes`` rather than ``labels``.
+  - :meth:`MultiIndex.set_labels` has been deprecated in favor of :meth:`MultiIndex.set_codes`
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: period at the end.

File: pandas/tests/indexes/multi/test_get_set.py [hunk 9]
```diff
@@ -232,102 +232,124 @@ def test_set_levels(idx):
...
+                                    inplace=True)
     assert inplace_return is None
-    assert_matching(ind2.labels, new_labels)
-    assert_matching(idx.labels, labels)
+    assert_matching(ind2.codes, new_codes)
+    assert_matching(idx.codes, codes)
 
     # label changing for levels of different magnitude of categories
     ind = pd.MultiIndex.from_tuples([(0, i) for i in range(130)])
+    new_codes = range(129, -1, -1)
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_get_set.py]: are these added tests?

**Review by jreback — CHANGES_REQUESTED** [review]
small doc comments. ping on green.

File: pandas/tests/indexes/multi/test_get_set.py (hunk 9 above)
Comment (topper-123) [reply pandas/tests/indexes/multi/test_get_set.py]: No, this is just git being a bit funky. This is just renamed code, but git displays it as new.

The real new tests are it below (line 325-341 ``test_set_labels_deprecated``). This is very similar code, but is testing that deprecation messages are emitted (i.e. FutureWarnings).

File: doc/source/whatsnew/v0.24.0.rst (hunk 6 above)
Comment (topper-123) [reply doc/source/whatsnew/v0.24.0.rst]: Not sure I understand.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @topper-123 nice change!

**Comment by jorisvandenbossche** [comment]
Thanks a lot for this @topper-123 !

**Comment by jorisvandenbossche** [comment]
There are still a bunch of warnings in the docs that should be fixed:

```
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4631
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:742: FutureWarning: .labels was deprecated in version 0.24.0. Use .codes instead.
  labels, = index.labels
<<<-------------------------------------------------------------------------
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4704
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:698: FutureWarning: .labels was deprecated in version 0.24.0. Use .codes instead.
  labels = getattr(columns, 'labels', None) or [
<<<-------------------------------------------------------------------------
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4704
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:725: FutureWarning: the 'labels' keyword is deprecated, use 'codes' instead
  return pd.MultiIndex(levels=new_levels, labels=labels, names=columns.names)
<<<-------------------------------------------------------------------------
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4704
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:742: FutureWarning: .labels was deprecated in version 0.24.0. Use .codes instead.
  labels, = index.labels
<<<-------------------------------------------------------------------------
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4768
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:698: FutureWarning: .labels was deprecated in version 0.24.0. Use .codes instead.
  labels = getattr(columns, 'labels', None) or [
<<<-------------------------------------------------------------------------
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4768
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:725: FutureWarning: the 'labels' keyword is deprecated, use 'codes' instead
  return pd.MultiIndex(levels=new_levels, labels=labels, names=columns.names)
<<<-------------------------------------------------------------------------
>>>-------------------------------------------------------------------------
Warning in /home/travis/build/pandas-dev/pandas/doc/source/io.rst at block ending on line 4768
Specify :okwarning: as an option in the ipython:: block to suppress this message
----------------------------------------------------------------------------
/home/travis/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pyarrow/pandas_compat.py:742: FutureWarning: .labels was deprecated in version 0.24.0. Use .codes instead.
  labels, = index.labels
<<<-------------------------------------------------------------------------
```

**Comment by jorisvandenbossche** [comment]
Ah, they all seem to come from parquet / feather. Will report upstream.

**Comment by jorisvandenbossche** [comment]
There are some other in the whatsnew files as well

**Pull request merged**